core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -215,10 +215,10 @@ private void doExecute(PooledChannel pooledChannel, ClientRequestContext ctx,
                     // If pipelining is disabled, return after the response is fully received.
                     final CompletableFuture<Void> completionFuture =
                             factory.useHttp1Pipelining() ? req.completionFuture() : res.completionFuture();
-                    completionFuture.handle((ret, cause) -> {
+                    completionFuture.handleAsync((ret, cause) -> {
                         pooledChannel.release();
                         return null;
-                    });
+                    }, ctx.eventLoop());
```

Comment:
Could we use direct invocation when the current thread is an event loop instead of always handling it asynchronously? For example, we could keep using `handle()` here and update the `PooledChannel.release()` implementation.

```java
@Override
public void release() {
    if (!eventLoop.inEventLoop()) {
        eventLoop.execute(() -> release0());
    } else {
        release0();
    }
}
```